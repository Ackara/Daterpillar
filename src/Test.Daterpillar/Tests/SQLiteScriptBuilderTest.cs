using ApprovalTests.Namers;
using ApprovalTests.Reporters;
using Acklann.Daterpillar.TextTransformation;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System;
using System.IO;
using Tests.Daterpillar.Constants;
using Tests.Daterpillar.Helpers;

namespace Test.Daterpillar.Tests
{
    [TestClass]
    [DeploymentItem(SampleData.Folder)]
    [DeploymentItem(KnownFile.DbConfig)]
    [DeploymentItem(KnownFile.x86SQLiteInterop)]
    [UseApprovalSubdirectory(nameof(ApprovalTests))]
    [UseReporter(typeof(DiffReporter), typeof(ClipboardReporter))]
    public class SQLiteScriptBuilderTest : DbTemplateBuilderTestBase
    {
        private static readonly string DataSource = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "datapillar-test.db3");

        [ClassCleanup]
        public static void Cleanup()
        {
            ApprovalTests.Maintenance.ApprovalMaintenance.CleanUpAbandonedFiles();
#if !DEBUG
            if (System.IO.File.Exists(DataSource)) System.IO.File.Delete(DataSource);
#endif
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Create_should_generate_a_sqlite_script_that_builds_a_new_schema_when_settings_are_enabled()
        {
            using (var connection = DatabaseHelper.CreateSQLiteConnection(DataSource))
            {
                var settings = new ScriptBuilderSettings()
                {
                    AppendScripts = true,
                    AppendComments = true,
                    CreateDatabase = true,
                    TruncateDatabaseIfItExist = true
                };

                RunSchemaTest<SQLiteScriptBuilder>(settings, connection);
            }
        }

        [TestMethod]
        [TestCategory(Trait.Integration)]
        public void Create_should_generate_a_sqlite_script_that_adds_a_new_column_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateSQLiteConnection(DataSource))
            {
                RunColumnTest<SQLiteScriptBuilder>(connection);
            }
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Create_should_generate_a_sqlite_script_that_adds_a_new_index_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateSQLiteConnection(DataSource))
            {
                RunIndexTest<SQLiteScriptBuilder>(connection);
            }
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Create_should_generate_a_sqlite_script_that_adds_a_new_foreign_key_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateSQLiteConnection(DataSource))
            {
                RunForeignKeyTest<SQLiteScriptBuilder>(connection);
            }
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Drop_should_generate_a_sqlite_script_that_removes_a_schema_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateSQLiteConnection(DataSource))
            {
                RunSchemaDropTest<SQLiteScriptBuilder>(connection);
            }
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Drop_should_generate_a_sqlite_script_that_removes_a_column_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateSQLiteConnection(DataSource))
            {
                RunColumnDropTest<SQLiteScriptBuilder>(connection);
            }
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Drop_should_generate_a_sqlite_script_that_removes_a_table_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateSQLiteConnection(DataSource))
            {
                RunTableDropTest<SQLiteScriptBuilder>(connection);
            }
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Drop_should_generate_a_sqlite_script_that_removes_a_index_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateSQLiteConnection(DataSource))
            {
                RunIndexDropTest<SQLiteScriptBuilder>(connection);
            }
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Drop_should_generate_a_sqlite_script_that_removes_a_foreign_key_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateSQLiteConnection(DataSource))
            {
                RunForeignKeyDropTest<SQLiteScriptBuilder>(connection);
            }
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Alter_should_generate_a_sqlite_modify_script_for_a_table_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateSQLiteConnection(DataSource))
            {
                RunAlterTableTest<SQLiteScriptBuilder>(connection);
            }
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Alter_should_generate_a_sqlite_modify_script_for_a_column_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateSQLiteConnection(DataSource))
            {
                RunAlterColumnTest<SQLiteScriptBuilder>(connection);
            }
        }
    }
}