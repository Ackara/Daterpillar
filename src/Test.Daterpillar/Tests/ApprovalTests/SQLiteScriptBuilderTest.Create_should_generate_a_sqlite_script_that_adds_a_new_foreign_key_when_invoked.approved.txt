/* ADD ([tbl1].[]) SCRIPT */
PRAGMA foreign_keys = 0;
CREATE TABLE [tbl1_temp_table] AS SELECT * FROM [tbl1];
DROP TABLE [tbl1];

CREATE TABLE IF NOT EXISTS [tbl1]
(
	[Id] INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
	[Name] VARCHAR(64) NOT NULL,
	[CategoryId] INTEGER NOT NULL,
	CONSTRAINT [tbl1_categoryid_to_tbl2_id_fkey1] FOREIGN KEY ([CategoryId]) REFERENCES [tbl2]([Id]) ON UPDATE CASCADE ON DELETE CASCADE
);

CREATE INDEX IF NOT EXISTS [cat_idx] ON [tbl1] ([CategoryId] ASC);

INSERT INTO [tbl1] ([Id], [Name], [CategoryId]) SELECT [Id], [Name], [CategoryId] FROM [tbl1_temp_table];
DROP TABLE [tbl1_temp_table];
PRAGMA foreign_keys = 1;
/* END SCRIPT */
